src/environment-logs-provider.ts

```diff
@@ -79,12 +79,9 @@ export class EnvironmentLogsProvider {
   }
 
   private async logCompletedDeployment(deploymentId: string) {
-    const steps = await apiClient.getDeploymentSteps(
-      deploymentId,
-      this.abortController
-    );
+    const stepsToLog = await this.getStepsToLog(deploymentId);
 
-    const stepsEvents = steps.map(async (step) => ({
+    const stepsEvents = stepsToLog.map(async (step) => ({
       name: step.name,
       events: await this.getStepLogs(deploymentId, step.name),
     }));
```

Comment:
@meniRoy since the deployment is already completed, why not getting all steps logs concurrently using `Promise.all`?

Reply:
We are fetching the logs for all steps concurrently; array.map doesn't wait for the previous promise to complete before it moves on to the next item. If I add a Promise.all here, it will slow down the process because the code will wait until all the logs are fetched before we start logging them. As you can see in the subsequent loop, we only await for the current step that we want to log, allowing the fetching of the logs for the next steps to continue in the background while logging.

Reply:
okay got you, yes I see that the logs are being fetched in next lines. But doesn't look like it's being fetched concurrently. If the deployment is already completed there is no reason to read the logs step by step.`

IIUC here the process `awaits` for each step one by one. Why not read all steps concurrently, and at the end just pipe the the message to the logger with $$ and ##

```
 for await (const step of stepsEvents) {
      this.log(`$$$ ${step.name}`);
      this.log("#".repeat(100));
      step.events.forEach((event) => this.log(stripAnsi(event.message)));
    }
```

Reply:
the `await` is on the entire array? on the iterated item ?

Reply:
I hope this will clarify that
`const getLogs = (stepId) => {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve(`logs for step ${stepId}`);
    }, 1000 * stepId);
  });
};
const log = async () => {
  const steps = [1, 2, 3, 4, 5];
  const logs = steps.map(async (stepId) => await getLogs(stepId));
  const start = Date.now();
  for await (const log of logs) {
    console.log(`${Date.now() - start}ms: ${log}`);
  }
  console.log(`${Date.now() - start}ms: done`);
};
log();
`
output:
`1002ms: logs for step 1
2001ms: logs for step 2
3001ms: logs for step 3
4001ms: logs for step 4
5000ms: logs for step 5
5000ms: done`

Reply:
I hope this will clarify that

```javascript
const getLogs = (stepId) => {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve(`logs for step ${stepId}`);
    }, 1000 * stepId);
  });
};
const log = async () => {
  const steps = [1, 2, 3, 4, 5];
  const logs = steps.map(async (stepId) => await getLogs(stepId));
  const start = Date.now();
  for await (const log of logs) {
    console.log(`${Date.now() - start}ms: ${log}`);
  }
  console.log(`${Date.now() - start}ms: done`);
};
log();
```
output:
1002ms: logs for step 1
2001ms: logs for step 2
3001ms: logs for step 3
4001ms: logs for step 4
5000ms: logs for step 5
5000ms: done

Reply:
it doesn't reflect a completed deployment. the logs are already available in the backend, you shouldn't wait for them.. In your example, each step ends in different time. In case of **live deployment** it makes sense to read the logs of each step , one by one.

Reply:
but it doesn't really matter currently. it's an optimization that can be applies at any time so let's move on

Reply:
we talked in the slack:

Logs are being fetched concurrently, but only when the next step Promise is resolved, the log is piped to the console.